pkg/controller/node/nodecontroller.go

```diff
@@ -601,6 +611,83 @@ func (nc *NodeController) Run() {
 	}()
 }
 
+// addOrRemoveNoScheduleTaintByCondition taint or remove NoSchedule taint on this node based on it's condition change.
+func (nc *NodeController) addOrRemoveNoScheduleTaintByCondition(node *v1.Node) (bool, bool) {
+	var changedToUnSchedulable, changedToSchedulable bool
+	observedMap := map[v1.NodeConditionType]*v1.NodeCondition{}
+	_, observedMap[v1.NodeReady] = v1.GetNodeCondition(&node.Status, v1.NodeReady)
+	_, observedMap[v1.NodeOutOfDisk] = v1.GetNodeCondition(&node.Status, v1.NodeOutOfDisk)
+	_, observedMap[v1.NodeNetworkUnavailable] = v1.GetNodeCondition(&node.Status, v1.NodeNetworkUnavailable)
+
+	// If there are saved node status, it only make sense to check conditions when some change happened.
+	if savedNodeStatus, found := nc.nodeStatusMap[node.GetName()]; found {
+		for conditionType, condition := range observedMap {
+			_, savedCondition := v1.GetNodeCondition(&savedNodeStatus.status, conditionType)
+			// something changed
+			if (condition != nil && savedCondition == nil) ||
+				(condition != nil && savedCondition != nil && condition.Status != savedCondition.Status) {
+				if unSchedulable(condition) {
+					changedToUnSchedulable = true
+					changedToSchedulable = false
+					// if one of the condition changed and becomes unschedulable, exit loop
+					break
+				} else {
+					changedToSchedulable = true
+				}
+			}
+			// if condition == nil && savedCondition != nil, we do nothing
+		}
+	} else {
+		// Otherwise, there are no saved node status, it always worth to check conditions
+		for _, condition := range observedMap {
+			if condition != nil && unSchedulable(condition) {
+				changedToUnSchedulable = true
+				changedToSchedulable = false
+				break
+			} else {
+				changedToSchedulable = true
+			}
+		}
+	}
+
+	if changedToUnSchedulable {
+		if err := nc.taintNodeNoSchedule(node); err != nil {
+			glog.Errorf("failed to add NoSchedule taint on node %v, %v", node.GetName(), err)
+		} else {
+			glog.V(2).Infof("Tainting Node %v with NoSchedule taint because its condition changed to unschedulable",
+				node.Name,
+			)
+		}
+	}
+
+	if changedToSchedulable {
+		if err := nc.removeNoScheduleTaint(node); err != nil {
+			glog.Errorf("failed to remove NoSchedule taint on node %v, %v", node.GetName(), err)
+		} else {
+			glog.V(2).Infof("Mark node %v as schedulable again because its condition changed to schedulable",
+				node.Name,
+			)
+		}
+	}
+
+	return changedToSchedulable, changedToUnSchedulable
+}
+
+func unSchedulable(cond *v1.NodeCondition) bool {
+	// We consider the node for scheduling only when its:
```

Comment:
I guess we could do
```
// isSchedulable is a flight check for node readiness, disk, and network.
```
as the header comment and delete the extra comments below since they're redundant to the implementation.